server/src/main/java/io/crate/execution/engine/aggregation/DocValueAggregator.java

```diff
@@ -40,5 +41,5 @@
     // that means there is always a final reduce step necessary
     // → never return final value, but always partial result
     @Nullable
-    public Object partialResult(T state);
+    public Object partialResult(RamAccounting ramAccounting, T state);
```

Comment:
I am not sure, but it seems that it is only the place so far where it would make sense to do the accounting.

Reply:
I think that's the only place in this context.

Reply:
I'm not sure if `partialResult` is a good fit.

`partialResult` is called at the end of the aggregation. So at the very end after the structures have been in memory already. I think `initialState` and `apply` (for the `collect` cases later on) would fit better.

This probably also has an influence on the abstractions used. And I second @kovrus comment that the avoidance of abstractions is what led to the performance improvements, so we have to be careful to re-introduce new abstractions (and re-run the benchmarks)

Reply:
> partialResult is called at the end of the aggregation. So at the very end after the structures have been in memory already. I think initialState and apply (for the collect cases later on) would fit better.

good point, I was somehow wrongly asuming that the structures were already accounted elsewhere.

Reply:
> And I second @kovrus comment that the avoidance of abstractions is what led to the performance improvements, so we have to be careful to re-introduce new abstractions (and re-run the benchmarks)

I totally agree that we should be careful in that direction. On the other hand we should also be careful with maybe too much (avoidable) concrete code.